src/bucket/BucketIndex.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "bucket/LedgerCmp.h"
+#include "lib/bloom_filter.hpp"
+#include "util/NonCopyable.h"
+#include <filesystem>
+#include <memory>
+#include <optional>
+#include <variant>
+
+namespace stellar
+{
+
+/**
+ * BucketIndex is an in-memory mapping of LedgerKey's to the file offset
+ * of the associated LedgerEntry in a given Bucket file. Because the set of
+ * LedgerKeys is too large to keep in memory, BucketIndex can either index
+ * individual keys or key ranges.
+ *
+ * For small buckets, an individual index is used for faster index lookup. For
+ * larger buckets, the range index is used. The range index cannot give an exact
+ * position for a given LedgerEntry or tell if it exists in the bucket, but can
+ * give an offset range for where the entry would be if it exists. Config flags
+ * determine the size of the range index, as well as what bucket size should use
+ * the individual index vs range index.
+ */
+
+class Config;
+
+struct BucketIndexRangeEntry
+{
+    LedgerKey lowerBound;
+    LedgerKey upperBound;
+
+    BucketIndexRangeEntry(LedgerKey low, LedgerKey high)
+        : lowerBound(low), upperBound(high)
+    {
+    }
+};
+
+// BucketIndex abstract interface
+class BucketIndex : public NonMovableOrCopyable
+{
+  public:
+    using RangeIndex = std::vector<BucketIndexRangeEntry>;
+    using IndividualIndex = std::vector<LedgerKey>;
+    using Iterator = std::variant<RangeIndex::const_iterator,
+                                  IndividualIndex::const_iterator>;
+
+    // Builds index for given bucketfile. This is expensive (> 20 seconds for
+    // the largest buckets) and should only be called once. If pageSize == 0 or
+    // if file size is less than the cutoff, individual key index is used.
+    // Otherwise range index is used, with the range defined by pageSize.
+    static std::unique_ptr<BucketIndex const>
+    createIndex(Config const& cfg, std::filesystem::path const& filename);
+
+    // Returns next page boundary > pos
+    static inline std::streamoff
+    nextPageBoundary(std::streamoff pos, std::streamoff pageSize)
+    {
+        return (pos & ~(pageSize - 1)) + pageSize;
```

Comment:
Yeah, originally it was due to low level disk IO optimizations. Those were all experimental though and mostly got cut out. Currently there's nothing that would break if the page size isn't a power of two (except for the nextPageBoundary calculation), but since we are reading the range index a page at a time it would be preferably if the pageSize was a multiple of the underlying disk sector size. I'm thinking I'll keep the assert and add comments.

Reply:
ok. Also, we should avoid duplicating the code here - we already have `BucketList::mask` that applies the same bitwise logic